File: Difficulty: Medium/K Closest Points to Origin/k-closest-points-to-origin.py
#{ 
 # Driver Code Starts
#Initial Template for Python 3
from typing import List


# } Driver Code Ends

class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        # Your code here
        for i in range(len(points)):
            #calculate square of distance from the origin
            dist = points[i][0] * points[i][0] + points[i][1] * points[i][1] 
            # append the square of distance from origin as third element to list for each point 
            points[i].append(dist)
        # sort the list w.r.t. square of distance from origin 
        points.sort(key=lambda x: x[2])

        # return first k points, priting of first two elements x,y is taken care of in the driver code itself
        return points[:k]
        

#{ 
 # Driver Code Starts.

if __name__ == "__main__":
    t = int(input())
    while t > 0:
        t -= 1
        k = int(input())
        n = int(input())
        points = []
        for _ in range(n):
            x, y = map(int, input().split())
            points.append([x, y])
        
        solution = Solution()
        ans = solution.kClosest(points, k)
        ans.sort()
        for point in ans:
            print(point[0], point[1])
        print("~")

# } Driver Code Ends

File: Difficulty: Medium/K Closest Points to Origin/k-closest-points-to-origin.cpp
class Solution {
  public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        // code here
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
        int n = points.size();
        for(int i = 0;i<n;i++){
            int first = points[i][0];
            int second = points[i][1];
            first = first*1ll*first;
            second = second*1ll*second;
            int sum = first + second;
            pq.push({sum,i});
        }
        vector<vector<int>> ans;
        while(!pq.empty()&&k>0){
            int req = pq.top().second;
            pq.pop();
            k--;
            ans.push_back(points[req]);
        }
        return ans;
    }
};